src/trace/implementations/merge_batcher_flat.rs

```diff
@@ -0,0 +1,325 @@
+            // Recycling buffer.
+            self.recycle(buffer, stash);
+        }
+        // Finish the kept data.
+        if !keep.is_empty() {
+            kept.push(keep);
+        }
+        if !ready.is_empty() {
+            readied.push(ready);
+        }
+    }
+
+    fn seal<B: Builder<Input = Self::Output, Time = Self::Time>>(
+        chain: &mut Vec<Self::Chunk>,
+        lower: AntichainRef<Self::Time>,
+        upper: AntichainRef<Self::Time>,
+        since: AntichainRef<Self::Time>,
+    ) -> B::Output {
+        let mut keys = 0;
+        let mut vals = 0;
+        let mut upds = 0;
+        {
+            let mut prev_keyval = None;
+            for buffer in chain.iter() {
+                for element in buffer.iter() {
+                    let (key, val, time, _) = FR::into_parts(element);
+                    if !upper.less_equal(&time) {
+                        if let Some((p_key, p_val)) = prev_keyval {
+                            if p_key != key {
+                                keys += 1;
+                                vals += 1;
+                            } else if p_val != val {
+                                vals += 1;
+                            }
+                        } else {
+                            keys += 1;
+                            vals += 1;
+                        }
+                        upds += 1;
+                        prev_keyval = Some((key, val));
+                    }
+                }
+            }
+        }
+        let mut builder = B::with_capacity(keys, vals, upds);
+        for mut chunk in chain.drain(..) {
+            builder.push(&mut chunk);
+        }
+
+        builder.done(lower.to_owned(), upper.to_owned(), since.to_owned())
+    }
+
+    fn account(chunk: &Self::Chunk) -> (usize, usize, usize, usize) {
+        let (mut size, mut capacity, mut allocations) = (0, 0, 0);
+        let cb = |siz, cap| {
+            size += siz;
+            capacity += cap;
+            allocations += 1;
+        };
+        chunk.heap_size(cb);
+        (chunk.len(), size, capacity, allocations)
+    }
+}
+
+struct FlatStackQueue<R: Region> {
+    list: FlatStack<R>,
+    head: usize,
+}
+
+impl<R: Region> Default for FlatStackQueue<R> {
+    fn default() -> Self {
+        Self::from(Default::default())
+    }
+}
+
+impl<R: Region> FlatStackQueue<R> {
+    fn pop(&mut self) -> R::ReadItem<'_> {
+        self.head += 1;
+        self.list.get(self.head - 1)
+    }
+
+    fn peek(&self) -> R::ReadItem<'_> {
+        self.list.get(self.head)
+    }
+
+    fn from(list: FlatStack<R>) -> Self {
+        FlatStackQueue { list, head: 0 }
+    }
+
+    fn done(self) -> FlatStack<R> {
+        self.list
+    }
+
+    fn is_empty(&self) -> bool {
+        self.head == self.list.len()
+    }
+
+    /// Return an iterator over the remaining elements.
+    fn iter(&self) -> impl Iterator<Item = R::ReadItem<'_>> + Clone {
+        self.list.iter().skip(self.head)
+    }
```

Comment:
This stuck out as potentially surprisingly expensive. Do we do this often, and when we do are we meant to drain? NB: re-read, and it seems like we call it at most twice for each buffer, once to size and then once to copy the elements. Anyhow, nothing wrong and we can discuss, but .. possibly expensive if skipping things isn't free.

Reply:
Good callout, should fix it: https://github.com/antiguru/flatcontainer/issues/36